### DISCUSSION TITLE
Idea of `LifespanManager` to have multiple lifespans in app

### DISCUSSION THREAD

**Discussion by uriyyo** [fastapi/fastapi#9397] [Show and tell]
The old approach with `on_startup`, `on_shutdown` is deprecated and will be removed in future versions of `starlette`.

The new `lifespan` approach has a small issue in that you can't have multiple `lifespans` defined. It's great to have a separate lifespan for db connection, cache connection, .etc.

I have created library called [fastapi-lifespan-manager](https://github.com/uriyyo/fastapi-lifespan-manager) which aimed to fixed this issue.

I am happy to open PR to add this feature to `fastapi` cause the library is pretty small itself.

Usage example:
```py
from typing import AsyncIterator

from fastapi import FastAPI
from pydantic import AnyUrl, BaseSettings
from redis.asyncio import Redis
from sqlalchemy.ext.asyncio import create_async_engine

from fastapi_lifespan_manager import LifespanManager, State


class Settings(BaseSettings):
    DB_URL: AnyUrl
    REDIS_DSN: AnyUrl | None = None


settings = Settings()

manager = LifespanManager()


# Lifespan functions to connect to the database
@manager.add
async def setup_db(app: FastAPI) -> AsyncIterator[State]:
    engine = await create_async_engine(settings.DB_URL)

    yield {"db": engine}

    await engine.dispose()


# Lifespan functions to connect to the cache, will be skipped if REDIS_DSN is None
if settings.REDIS_DSN:
    @manager.add
    async def setup_cache(app: FastAPI) -> AsyncIterator[State]:
        redis = await Redis.from_url(settings.REDIS_DSN)

        yield {"cache": redis}

        await redis.close()
        await redis.wait_closed()

app = FastAPI(lifespan=manager)

```

**Comment by notypecheck** [comment]
@uriyyo You could use this
```py
import contextlib
from collections.abc import AsyncIterator, Callable, Sequence
from contextlib import AbstractAsyncContextManager

from fastapi import FastAPI


@contextlib.asynccontextmanager
async def _manager(
    app: FastAPI,
    lifespans: Sequence[Callable[[FastAPI], AbstractAsyncContextManager[None]]],
) -> AsyncIterator[None]:
    exit_stack = contextlib.AsyncExitStack()
    async with exit_stack:
        for lifespan in lifespans:
            await exit_stack.enter_async_context(lifespan(app))
        yield


class Lifespans:
    def __init__(
        self,
        lifespans: Sequence[Callable[[FastAPI], AbstractAsyncContextManager[None]]],
    ) -> None:
        self.lifespans = lifespans

    def __call__(self, app: FastAPI) -> AbstractAsyncContextManager[None]:
        self.app = app
        return _manager(app, lifespans=self.lifespans)


@contextlib.asynccontextmanager
async def one(app: FastAPI) -> AsyncIterator[None]:
    print("One")
    yield
    print("End One")


@contextlib.asynccontextmanager
async def two(app: FastAPI) -> AsyncIterator[None]:
    print("Two")
    yield
    print("End Two")


app = FastAPI(lifespan=Lifespans([one, two]))
```

**Comment by beornharris** [comment]
Thanks both for your help and inspiration!  I wouldnt have found my solution without you...
I would like to suggest a (maybe) simpler approach though

```
from contextlib import asynccontextmanager, AsyncExitStack
from typing import List
from fastapi import FastAPI, APIRouter

def app_lifespan(lifespans: List):
    @asynccontextmanager
    async def _lifespan_manager(app: FastAPI):
        exit_stack = AsyncExitStack()
        async with exit_stack:
            for lifespan in lifespans:
                await exit_stack.enter_async_context(lifespan(app))
            yield
            
    return _lifespan_manager

@asynccontextmanager
async def one(app: FastAPI) -> AsyncIterator[None]:
    print("One")
    yield
    print("End One")


@asynccontextmanager
async def two(app: FastAPI) -> AsyncIterator[None]:
    print("Two")
    yield
    print("End Two")


app = FastAPI(lifespan=app_lifespan([one, two]))
```

I confess I did not run this example, but the code is ripped from a working version that I am using....so apologies for any oversights.
I hope this is useful/helpful to someone.

**Comment by boludoz** [comment]
For good reason, it's not possible. It's been known since the 1970s that Model-View-Controller (MVC) and its derivatives rule. Build the database externally. It's an unorthodox solution, better start a task. 

```
async def task_hola():
    """Tarea que imprime un mensaje cada 5 segundos"""
    while True:
        print("👋 Hola desde el lifespan")
        code_queue.put("👋 Hola desde el lifespan")
        await asyncio.sleep(5)

async def task_hola_dos():
    """Tarea que imprime un mensaje cada 10 segundos"""
    while True:
        print("👋👋 Hola desde el lifespan")
        code_queue.put("👋👋 Hola desde el lifespan")
        await asyncio.sleep(10)

@asynccontextmanager
async def while_hola(app: FastAPI):
    """Lifespan que ejecuta una tarea periódica"""
    task = asyncio.create_task(task_hola())
    task_dos = asyncio.create_task(task_hola_dos())
    print("👋 Tarea periódica iniciada")
    try:
        yield
    finally:
        task.cancel()
        task_dos.cancel()
        print("🛑 Tarea periódica cancelada")

```

**Reply by beornharris** [reply]
No - I think you missed the point.  In your solution, you lose the benefits of a lifespan (properly cleaning up your code etc).
In any case, "Lifespans" are already a thing - this question was around managing multiple unrelated Lifespans.
Thanks for contributing though :)

**Reply by boludoz** [reply]
> No - I think you missed the point. In your solution, you lose the benefits of a lifespan (properly cleaning up your code etc). In any case, "Lifespans" are already a thing - this question was around managing multiple unrelated Lifespans. Thanks for contributing though :)

I think discrediting my answer with that argument is pretty lame. The "Finally" statement is RIIA programming. It makes sure to properly clean up resources when the program ends. Also, leave that resource cleanup to C and low-level languages. This is Python, a language designed for people who aren't that familiar with garbage collection and such things, which are practically obsolete today.

**Reply by boludoz** [reply]
Let's start from the premise that this isn't correct or necessary. If you do this, depending on your code, it will be slow. Await isn't synonymous with multicore (or even threading, in principle), so you're sharing processor time with your database. It's better and cleaner to use separate processes and communicate between them through an API with an asynchronous client like HTTPX, which is precisely what FastAPI is designed for.

**Reply by notypecheck** [reply]
@boludoz If objects are eventually cleaned up (not in all python implementations, I think) it's not a reason to not manually clean up resources you've used, if you want to periodically run a task in background using `asyncio.Event` to stop it would be preferential I think:
```py
import asyncio
import contextlib


async def run_until_stopped(event: asyncio.Event, interval: float) -> None:
    wait_task = asyncio.create_task(event.wait())
    while True:
        sleep_task = asyncio.create_task(asyncio.sleep(interval))
        await asyncio.wait([wait_task, sleep_task], return_when=asyncio.FIRST_COMPLETED)

        if wait_task.done():
            sleep_task.cancel()
            return
        print("Task runs")


@contextlib.asynccontextmanager
async def lifespan():
    stopped = asyncio.Event()
    async with asyncio.TaskGroup() as tg:
        tg.create_task(run_until_stopped(event=stopped, interval=0.1))
        print("Before yield")
        yield
        stopped.set()
        print("After yield")


async def main() -> None:
    async with lifespan():
        await asyncio.sleep(1)


asyncio.run(main())
```
You may want to choose to just call `Task.cancel` when using a task group too, but if you gracefully shutdown tasks like this you ensure that they don't cancel mid execution.

**Reply by boludoz** [reply]
Well, I've tried to clean it up, that command will be executed at the end of the program, it's just an example.

**Comment by rgov** [comment]
The solutions above discuss multiple lifespan functions at the top level. But for mounted sub-applications, I've found that lifespan handlers are only called on the top-level application. In my case, I would like two sub-applications mounted at `/v1` and `/v2`, each with their own lifespan logic.

You could imagine the top-level lifespan handler might look like this, with a `cascade_lifespan()` returning a context manager that sends `lifespan.startup` to our child apps as we enter it and `lifespan.shutdown` when we exit:

```python
@contextlib.asynccontextmanager
async def top_lifespan(app):
    async with cascade_lifespan(app):
        # now all of our children have already started up
        yield
        # now all of our children are going to shut down
```

A simple version looks like this:

```python
@contextlib.asynccontextmanager
async def cascade_lifespan(parent):
    async with contextlib.AsyncExitStack() as stack:
        for route in parent.router.routes:
            if isinstance(route, fastapi.routing.Mount):
                await stack.enter_async_context(
                    route.app.router.lifespan_context(route.app)
                )
        yield    
```

But this is unfortunately *too* simple and breaks down:

1. It assumes every mount is a `starlette.routing.Router` and relies on implementation details of this class. But it can be any ASGI application callable.

    What we really want to do is have access to inject events into each callable's `receive()` queue. 

3. Per the spec, an application may raise an exception:

    > If an exception is raised when calling the application callable with a `lifespan.startup` message or a `scope` with type `lifespan`, the server must continue but not send any lifespan events.

    This means one child's exception must not stop the other children, and we need to avoid sending `lifespan.shutdown` later on.

    1. Since we are calling `lifespan_context()` directly, any exceptions that *do* get raised will be caught by the top-level `starlette.routing.Router.lifespan()` function that called us, which probably confuses things more.

4. The child application's lifespan handler can yield a `state` dictionary that should be passed to later requests. This is just dropped by this dummy implementation.

    It would be nice, maybe, if the state were 'namespaced' to the application, so that the app's request handler only receives its own state. However the spec specifically says the ASGI server can't meddle with the state dictionary, and that "applications should be cooperative by properly naming their keys such that they will not collide with other frameworks or middleware."

<details>
<summary>Here is a slightly more complete, but pretty convoluted implementation. It creates a task for each app callable to send it lifespan events. This version also doesn't handle exceptions or state correctly.</summary>

```python
@contextlib.asynccontextmanager
async def memory_stream_context(buffer_size):
    send, receive = anyio.create_memory_object_stream(buffer_size)
    try:
        yield send, receive
    finally:
        await send.aclose()
        await receive.aclose()


def _collect_subapps(app):
    for r in getattr(app, 'routes', ()):
        if isinstance(r, fastapi.routing.Mount):
            yield from _collect_subapps(r.app)  # comment out if you don't want recursive enumeration
            yield r.app


def _lifespan_driver(app, tg):
    @contextlib.asynccontextmanager
    async def _cm():
        scope = {
            'type': 'lifespan',
            'asgi': {'version': '3.0', 'spec_version': '1.0'},
            'state': {},
        }

        async with (
            memory_stream_context(0) as (
                drv_send,
                app_recv,
            ),
            memory_stream_context(0) as (
                app_send,
                drv_recv,
            ),
        ):
            tg.start_soon(app, scope, app_recv.receive, app_send.send)

            # startup
            await drv_send.send({'type': 'lifespan.startup'})
            if (msg := await drv_recv.receive())[
                'type'
            ] != 'lifespan.startup.complete':
                raise RuntimeError(f'startup failed: {msg}')

            try:
                yield
            finally:
                # shutdown
                await drv_send.send({'type': 'lifespan.shutdown'})
                if (msg := await drv_recv.receive())[
                    'type'
                ] != 'lifespan.shutdown.complete':
                    raise RuntimeError(f'shutdown failed: {msg}')

    return _cm()

@contextlib.asynccontextmanager
async def cascade_lifespan(root_app):
    async with (
        anyio.create_task_group() as tg,
        contextlib.AsyncExitStack() as stack,
    ):
        for sub in _collect_subapps(root_app):  # inner → outer
            await stack.enter_async_context(_lifespan_driver(sub, tg))
        yield
```
</details>

**Comment by CharlieReitzel** [comment]
Or ... undeprecate the original event handlers which are, anyway, a cleaner syntax.

The stated reason for deprecation is that global (or, at least, module level) variables are required to coordinate startup/shutdown work.  This is not a compelling rationale when the service-level features typically access singleton objects (as shown in all the examples!).  Especially in a language with weak member access controls like Python.

In short, it wasn't broken, why non un-fix it?
